tests/Moq.Tests/InvocationsFixture.cs

```diff
@@ -45,11 +47,131 @@ public void MockInvocationsIncludeArguments()
 		}
 
+		[Fact]
+		public void MockInvocationsIncludeReturnValue_NoSetup()
+		{
+			var mock = new Mock<IComparable>();
+
+			var obj = new object();
+
+			mock.Object.CompareTo(obj);
+
+			var invocation = mock.Invocations[0];
+
+			Assert.Equal(0, invocation.ReturnValue);
+			Assert.Null(invocation.Exception);
+		}
+
+		[Fact]
+		public void MockInvocationsIncludeReturnValue_Setup()
+		{
+			var mock = new Mock<IComparable>();
+			var obj = new object();
+			mock.Setup(c => c.CompareTo(obj)).Returns(42);
+
+			mock.Object.CompareTo(obj);
+
+			var invocation = mock.Invocations[0];
+
+			Assert.Equal(42, invocation.ReturnValue);
+			Assert.Null(invocation.Exception);
+		}
+
+		[Fact]
+		public void MockInvocationsIncludeReturnValue_BaseCall()
+		{
+			var mock = new Mock<Random>(1) // seed: 1
+			{
+				CallBase = true,
+			};
+
+			mock.Object.Next();
+
+			var invocation = mock.Invocations[0];
+
+			Assert.Equal(new Random(Seed: 1).Next(), invocation.ReturnValue);
+			Assert.Null(invocation.Exception);
+		}
+
+		[Fact]
+		public void MockInvocationsIncludeReturnValue_ReturnsException()
+		{
+			var mock = new Mock<ICloneable>();
+			var returnValue = new Exception();
+			mock.Setup(c => c.Clone()).Returns(returnValue);
+
+			mock.Object.Clone();
+
+			var invocation = mock.Invocations[0];
+
+			Assert.Equal(returnValue, invocation.ReturnValue);
+			Assert.Null(invocation.Exception);
+		}
+
+		[Fact]
+		public void MockInvocationsIncludeException_Setup()
+		{
+			var mock = new Mock<IComparable>();
+			var exception = new Exception("Message");
+			mock.Setup(c => c.CompareTo(It.IsAny<object>())).Throws(exception);
+
+			var thrown = Assert.Throws<Exception>(() => mock.Object.CompareTo(null));
+			
+			Assert.Equal(exception.Message, thrown.Message);
+
+			var invocation = mock.Invocations[0];
+			Assert.Same(thrown, invocation.Exception);
+		}
+
+		[Fact]
+		public void MockInvocationsIncludeException_BaseCall_Nonvirtual()
+		{
+			var mock = new Mock<List<int>>()
+			{
+				CallBase = true,
+			};
+
+			Assert.Throws<ArgumentException>(() => mock.Object.GetRange(1, 1));
+			// Base-call exceptions are not recorded
+			Assert.Empty(mock.Invocations);
+		}
+
+		[Fact]
+		public void MockInvocationsIncludeException_BaseCall_Virtual()
+		{
+			var mock = new Mock<Test>()
+			{
+				CallBase = true,
+			};
+
+			var thrown = Assert.Throws<InvalidOperationException>(() => mock.Object.ThrowingVirtualMethod());
+
+			Assert.Equal("Message", thrown.Message);
+
+			var invocation = mock.Invocations[0];
+			Assert.Same(thrown, invocation.Exception);
+		}
+
+		[Fact]
+		public void MockInvocationsIncludeException_MockException()
+		{
+			var mock = new Mock<ICloneable>(MockBehavior.Strict);
+
+			var thrown = Assert.Throws<MockException>(() => mock.Object.Clone());
+
+			Assert.Equal(
+@"ICloneable.Clone() invocation failed with mock behavior Strict.
+All invocations on the mock must have a corresponding setup.", thrown.Message);
```

Comment:
All good. If you wanted to simplify this test code a little further, you could query `thrown.Reasons == MockExceptionReasons.NoSetup` instead of comparing the message. That'll make the test less sensitive to potential future changes to that message's precise wording. (`MockException.Reasons` is an internal property that exists specifically for our test suite.)